runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/TimerReference.java

```diff
@@ -27,32 +27,23 @@
  */
 @AutoValue
 public abstract class TimerReference {
+
   /** Create a timer reference. */
-  public static TimerReference of(
-      PipelineNode.PTransformNode transform,
-      String localName,
-      PipelineNode.PCollectionNode collection) {
-    return new AutoValue_TimerReference(transform, localName, collection);
+  public static TimerReference of(PipelineNode.PTransformNode transform, String localName) {
+    return new AutoValue_TimerReference(transform, localName);
   }
 
   /** Create a timer reference from a TimerId proto and components. */
   public static TimerReference fromTimerId(
       RunnerApi.ExecutableStagePayload.TimerId timerId, RunnerApi.Components components) {
     String transformId = timerId.getTransformId();
     String localName = timerId.getLocalName();
-    String collectionId = components.getTransformsOrThrow(transformId).getInputsOrThrow(localName);
     RunnerApi.PTransform transform = components.getTransformsOrThrow(transformId);
-    RunnerApi.PCollection collection = components.getPcollectionsOrThrow(collectionId);
-    return of(
-        PipelineNode.pTransform(transformId, transform),
-        localName,
-        PipelineNode.pCollection(collectionId, collection));
+    return of(PipelineNode.pTransform(transformId, transform), localName);
   }
 
   /** The PTransform that uses this timer. */
   public abstract PipelineNode.PTransformNode transform();
   /** The local name the referencing PTransform uses to refer to this timer. */
   public abstract String localName();
-  /** The PCollection that backs this timer. */
-  public abstract PipelineNode.PCollectionNode collection();
```

Comment:
I don't plan on changing the representation at runtime now, but perhaps it will be in the future. For now, you can look up the collection in the inputs and outputs dict of a process bundle descriptor using the local name.

Reply:
Ok, found it, for example:

```java
stageBundleFactory
.getProcessBundleDescriptor()
// yes this is required twice
.getProcessBundleDescriptor()
.getPcollectionsMap()
.get("ParDo(Anonymous)2/ParMultiDo(Anonymous).foo.out:0")
.getUniqueName()
```
=> `"ParDo(Anonymous)2/ParMultiDo(Anonymous).foo"`

Reply:
You should be able to get the name of the PCollection from `ref.transform().getTransform().getInputsOrThrow(ref.localName())` as well.

Reply:
Actually, I don't have the local timer name available, only the timer pcollection id. If we move TimerReference#collection, I will have to remove ".out:0" from the input collection id, to look up the corresponding TimerReference.

Reply:
Maybe I'm missing something here, otherwise I'd keep the collection method.

Reply:
Could you provide a pointer to the code you're trying to use this from?

Reply:
Here I use the `collection()` method: https://github.com/apache/beam/blob/68d1f40ad658ff7a71fb3de32e48c2ef1cdd22cc/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java#L462

To resolve the Timer here: https://github.com/apache/beam/blob/68d1f40ad658ff7a71fb3de32e48c2ef1cdd22cc/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java#L576

Reply:
FYI, I've solved it like this: https://github.com/apache/beam/pull/6981/commits/49ea0b96825ccd8e461bfa94946ae7c21799ed6b